base/structure.py

```diff
@@ -1314,14 +1546,21 @@ def recurse(offset, available):
     @utils.string.decorate_arguments('name')
     def add(self, name):
         '''Append the specified member `name` with the default type at the end of the structure.'''
-        offset = self.owner.size + self.baseoffset
-        return self.add(name, int, offset)
+        return self.add(name, int)
     @utils.multicase(name=(six.string_types, tuple))
     @utils.string.decorate_arguments('name')
     def add(self, name, type):
         '''Append the specified member `name` with the given `type` at the end of the structure.'''
+
+        # If this structure is a union, then the offset is always 0.
+        if is_union(self.ptr):
+            return self.add(name, type, 0)
```

Comment:
This might need to be "self.baseoffset" instead of 0 due to the way it's used in the code that follows.